src/python/pants/jvm/strip_jar/strip_jar.py

```diff
@@ -0,0 +1,201 @@
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, Directory, FileContent
+from pants.engine.internals.native_engine import MergeDigests, RemovePrefix
+from pants.engine.process import FallibleProcessResult, ProcessExecutionFailure, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import InternalJdk, JvmProcess
+from pants.jvm.resolve.coursier_fetch import ToolClasspath, ToolClasspathRequest
+from pants.jvm.resolve.jvm_tool import GenerateJvmLockfileFromTool
+from pants.option.global_options import ProcessCleanupOption
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet
+
+_STRIP_JAR_BASENAME = "StripJar.java"
+_OUTPUT_PATH = "__stripped_jars"
+
+
+class StripJarToolLockfileSentinel(GenerateToolLockfileSentinel):
+    resolve_name = "strip-jar"
+
+
+@dataclass(frozen=True)
+class StripJarRequest:
+    digest: Digest
+    filenames: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class FallibleStripJarResult:
+    process_result: FallibleProcessResult
+
+
+@dataclass(frozen=True)
+class StripJarCompiledClassfiles:
+    digest: Digest
+
+
+@rule(level=LogLevel.DEBUG)
+async def strip_jar(
+    processor_classfiles: StripJarCompiledClassfiles, jdk: InternalJdk, request: StripJarRequest
+) -> FallibleStripJarResult:
+
+    filenames = list(request.filenames)
+
+    input_path = "__jars_to_strip"
+    toolcp_relpath = "__toolcp"
+    processorcp_relpath = "__processorcp"
+
+    lockfile_request = await Get(GenerateJvmLockfileFromTool, StripJarToolLockfileSentinel())
+
+    tool_classpath, prefixed_jars_digest = await MultiGet(
+        Get(
+            ToolClasspath,
+            ToolClasspathRequest(lockfile=lockfile_request),
+        ),
+        Get(Digest, AddPrefix(request.digest, input_path)),
+    )
+
+    extra_immutable_input_digests = {
+        toolcp_relpath: tool_classpath.digest,
+        processorcp_relpath: processor_classfiles.digest,
+    }
+
+    process_result = await Get(
+        FallibleProcessResult,
+        JvmProcess(
+            jdk=jdk,
+            classpath_entries=[
+                *tool_classpath.classpath_entries(toolcp_relpath),
+                processorcp_relpath,
+            ],
+            argv=["org.pantsbuild.stripjar.StripJar", input_path, _OUTPUT_PATH, *filenames],
+            input_digest=prefixed_jars_digest,
+            extra_immutable_input_digests=extra_immutable_input_digests,
+            output_directories=(_OUTPUT_PATH,),
+            extra_nailgun_keys=extra_immutable_input_digests,
+            description=f"Stripping jar {filenames[0]}",
+            level=LogLevel.DEBUG,
+        ),
+    )
+
+    return FallibleStripJarResult(process_result=process_result)
+
+
+@rule(level=LogLevel.DEBUG)
+async def resolve_fallible_result_to_strip_jar(
+    fallible_result: FallibleStripJarResult,
+    process_cleanup: ProcessCleanupOption,
+) -> Digest:
```

Comment:
This indirection via `FallibleStripJarResult` is probably not necessary: assuming that the `description` for the process is good, roughly the same error message would be rendered by requesting `ProcessResult` rather than `FallibleProcessResult`.

Reply:
removed.